ChiModules/KEigenvalueSolver/IterativeMethods/power_iteration.cc

```diff
@@ -18,168 +18,127 @@ typedef sweep_namespace::SchedulingAlgorithm SchedulingAlgorithm;
-
-      if (si_nit==0) rho = 0.0;
-      if (pw_change<std::max(groupset.residual_tolerance*rho,1.0e-10))
-        si_converged = true;
-
-      // ----- Print iteration information
-      std::string offset = "    ";
-      std::stringstream si_iter_info;
-      si_iter_info
-        << chi_program_timer.GetTimeString() << " "
-        << offset
-        << "WGS groups ["
-        << groupset.groups.front().id
-        << "-"
-        << groupset.groups.back().id
-        << "]"
-        << " Source Iteration " << std::setw(5) << si_nit
-        << " Point-wise change " << std::setw(14) << pw_change;
-        
-      if (si_converged)
-        si_iter_info << " CONVERGED\n";
-      chi_log.Log(LOG_0VERBOSE_1) << si_iter_info.str();
-
-      if (si_converged) {
-        chi_log.Log(LOG_0VERBOSE_1)
-            << "\nSource iterations converged in "
-            << si_nit << " iteratrions.\n";
-        break;
+      ComputeSweepOrderings(groupset);
+      InitFluxDataStructures(groupset);
+
+      InitWGDSA(groupset);
+      InitTGDSA(groupset);
+
+      groupset.angle_agg.ZeroIncomingDelayedPsi();
+
+      //======================================== Setup sweep chunk
+      auto sweep_chunk = SetSweepChunk(groupset);
+      MainSweepScheduler sweep_scheduler(SchedulingAlgorithm::DEPTH_OF_GRAPH,
+                                         groupset.angle_agg,
+                                         *sweep_chunk);
+
+      //======================================== Tool the sweep chunk
+      sweep_scheduler.sweep_chunk.SetDestinationPhi(phi_new_local);
+
+      //======================================== Precompute the fission source
+      q_moments_local.assign(q_moments_local.size(), 0.0);
+      SetSource(groupset, q_moments_local,
+                APPLY_AGS_FISSION_SOURCE |
+                APPLY_WGS_FISSION_SOURCE);
+
+      //normalize q by k_eff
+      for (auto& q : q_moments_local) q /= k_eff;
+
+      //======================================== Converge the scattering source
+      //                                         with a fixed fission source
+      if (groupset.iterative_method == IterativeMethod::CLASSICRICHARDSON)
+      {
+        ClassicRichardson(groupset, gs, sweep_scheduler,
+                          APPLY_WGS_SCATTER_SOURCE |
+                          APPLY_AGS_SCATTER_SOURCE,
+                          false);
+      }
+      else if (groupset.iterative_method == IterativeMethod::GMRES)
+      {
+        GMRES(groupset, gs, sweep_scheduler,
+              APPLY_WGS_SCATTER_SOURCE,
+              APPLY_AGS_SCATTER_SOURCE,
+              false);
       }
-    }//for source iterations
 
-    if (!si_converged) {
-      chi_log.Log(LOG_ALLVERBOSE_1)
-          << "\n!!!WARNING!!! "
-             "Source iterations did not converge.\n";
-    }
+      CleanUpWGDSA(groupset);
+      CleanUpTGDSA(groupset);
 
-    // ----- Recompute eigenvalue
+      ResetSweepOrderings(groupset);
+
+      MPI_Barrier(MPI_COMM_WORLD);
```

Comment:
Is this barrier necessary?  If so, why?

Reply:
Please also do not remove this one. Same reasons as above. I have never identified these barriers to take any appreciable amount of time to clear with any simulation other than a really tiny one and even then the time taken is very small because the num_procs is also low. I have yet to be proven wrong.

Reply:
I can promise that unnecessary synchronization points run amok will eventually bite us.  Our jobs so far have been modest at best. At 1M+ MPI processes (or less depending on the interconnect), performance will suffer greatly.